pkg/autodiscovery/providers/cloudfoundry.go

```diff
@@ -0,0 +1,208 @@
+func (cf CloudFoundryConfigProvider) Collect() ([]integration.Config, error) {
+	cf.bbsCache.RLock()
+	defer cf.bbsCache.RUnlock()
+
+	cf.lastCollected = time.Now()
+	desiredLRPs := cf.bbsCache.GetDesiredLRPs()
+	allConfigs := []integration.Config{}
+	for _, desiredLRP := range desiredLRPs {
+		newConfigs := cf.getConfigsFromDesiredLRP(desiredLRP)
+		log.Debugf("Successfully got %d configs for app %s", len(newConfigs), desiredLRP.AppGUID)
+		allConfigs = append(allConfigs, newConfigs...)
+	}
+	return allConfigs, nil
+}
+
+func (cf CloudFoundryConfigProvider) getConfigsFromDesiredLRP(desiredLRP cloudfoundry.DesiredLRP) []integration.Config {
+	allConfigs := []integration.Config{}
+
+	for adName, adVal := range desiredLRP.EnvAD {
+		// initially, let's assume a non-container service; we'll change to container service in
+		// `expandPerContainerChecks` if necessary
+		id := cloudfoundry.NewADNonContainerIdentifier(desiredLRP, adName)
+		// we need to convert adVal to map[string]string to pass it to extractTemplatesFromMap
+		convertedADVal := map[string]string{}
+		for k, v := range adVal {
+			convertedADVal[k] = string(v)
+		}
+		parsedConfigs, errs := extractTemplatesFromMap(id.String(), convertedADVal, "")
+		for _, err := range errs {
+			log.Errorf("Cannot parse endpoint template for service %s of app %s: %s, skipping",
+				adName, desiredLRP.AppGUID, err)
+		}
+
+		vcVal, vcOk := desiredLRP.EnvVcapServices[adName]
+		variables, varsOk := adVal["variables"]
+		success := false
+		if vcOk {
+			// if service is found in VCAP_SERVICES (non-container service), we will run a single check per App
+			err := cf.renderExtractedConfigs(parsedConfigs, variables, vcVal)
+			cf.assignNodeNameToNonContainerChecks(parsedConfigs, desiredLRP)
+			if err != nil {
+				log.Errorf("Failed to render config for service %s of app %s: %s", adName, desiredLRP.AppGUID, err)
+			} else {
+				success = true
+			}
+		} else if varsOk {
+			log.Errorf("Service %s for app %s has variables configured, but is not present in VCAP_SERVICES", adName, desiredLRP.AppGUID)
+		} else {
+			// if a service is not in VCAP_SERVICES and has no "variables" configured, we want to run a check per container
+			parsedConfigs = cf.expandPerContainerChecks(parsedConfigs, desiredLRP, adName)
+			success = true
+		}
+		if success {
+			// mark all checks as cluster checks
+			for i := range parsedConfigs {
+				parsedConfigs[i].ClusterCheck = true
+			}
+			allConfigs = append(allConfigs, parsedConfigs...)
+		}
+	}
+
+	return allConfigs
+}
+
+func (cf CloudFoundryConfigProvider) assignNodeNameToNonContainerChecks(configs []integration.Config, desiredLRP cloudfoundry.DesiredLRP) {
+	aLRPs := cf.bbsCache.GetActualLRPsFor(desiredLRP.AppGUID)
+
+	if len(aLRPs) > 0 {
+		aLRP := aLRPs[0]
+		log.Debugf("All non-container checks for app %s will run on Cell %s", desiredLRP.AppGUID, aLRP.CellID)
+		for i := range configs {
+			configs[i].NodeName = aLRP.CellID
```

Comment:
Clarification: in the cluster checks dispatcher these will be considered endpoint checks - configs with empty `NodeName` field will be considered cluster checks.
The two types of checks are exposed via two different endpoints and each type requires a specific client (config provider on node agents). Endpoint checks are consumed by the `endpointschecks` config provider while cluster checks are consumed by `clusterchecks`.

Reply:
Thanks for the clarification. I think this is fine as long as we can have node agents configured to consume both endpoints checks and cluster checks. Is that possible?

Reply:
yes - it's the case in kuberentes.